native/acter/src/api/account/three_pid.rs

```diff
@@ -93,27 +145,90 @@ impl ThreePidManager {
                 };
                 account.set_account_data(content).await?;
 
+                Ok(ThreePidEmailTokenResponse {
+                    inner,
+                    client_secret,
+                })
+            })
+            .await?
+    }
+
+    // add 3pid on the homeserver for this account
+    // this 3pid may be used by the homeserver to authenticate the user during sensitive operations
+    #[cfg(feature = "testing")]
+    pub async fn add_3pid(
+        &self,
+        client_secret: String,
+        sid: String,
+        password: String,
+    ) -> Result<bool> {
+        let client = self.client.deref().clone();
+        let account = self.account.clone();
+        let user_id = self.user_id.clone();
+        let client_secret = ClientSecret::parse(&client_secret)?;
+        let sid = SessionId::parse(&sid)?; // it was already related with email or msisdn
+
+        RUNTIME
+            .spawn(async move {
+                let capabilities = client.get_capabilities().await?;
+                if !capabilities.thirdparty_id_changes.enabled {
+                    bail!("Server doesn't support 3pid change");
+                }
+                if let Err(e) = account.add_3pid(&client_secret, &sid, None).await {
+                    let Some(inf) = e.as_uiaa_response() else {
+                        return Err(clearify_error(e));
+                    };
+                    let pswd = assign!(Password::new(user_id.into(), password), {
+                        session: inf.session.clone(),
+                    });
+                    let auth_data = AuthData::Password(pswd);
+                    account
+                        .add_3pid(&client_secret, &sid, Some(auth_data))
+                        .await
+                        .map_err(clearify_error)?;
+                }
                 Ok(true)
             })
             .await?
     }
 
+    pub async fn get_3pids(&self, address: String) -> Result<Vec<ThreePid>> {
+        let account = self.account.clone();
+
+        RUNTIME
+            .spawn(async move {
+                let resp = account.get_3pids().await.map_err(clearify_error)?;
+                let records = resp
+                    .threepids
+                    .iter()
+                    .map(|x| ThreePid {
+                        address: x.address.clone(),
+                        medium: x.medium.clone(),
+                        added_at: x.added_at,
+                        validated_at: x.validated_at,
+                    })
```

Comment:
would be nicer if we added `::new(X)` on `ThreePid` so this would just be:
```suggestion
                    .into_iter()
                    .map(ThreePid::new)
```

Reply:
Yeah, good idea.
Fixed by https://github.com/acterglobal/a3/pull/1896/commits/c153ca45fb9a66c99188912c59de6e477597d812